### PR TITLE
Tidy up default connection flows

### PR DESCRIPTION
**Pull request by joshuahuburn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
https://improbableio.atlassian.net/browse/UNR-965

This PR removes a lot of unnecessary code from `SpatialGameInstance` as we no longer need the hacks to launch a net driver.
We now make use of Unreal's default connection flow which is to create a net driver and auto connect to a server when a Host is specified as the first commandline argument.
If there is no host present then the clients will launch offline. Please see the following doc for how default connection works in the GDK: https://docs.google.com/spreadsheets/d/1dLo-4E7rEetmNnHN1LCjFpYrN0Pzo3ist7GHqC2kC6U/edit#gid=0

Since the SpatialOS Launcher does not specify a host in the same way as native Unreal, we have a check for a locator loginToken which will force a netdriver and thus allow automatic connection.
Since developers may want to use the Launcher but start in an offline mode, there is now a `PreventAutoConnectWithLocator` flag which can be set.

#### Release note
Bugfix: Fixed up default connection flows.

#### Tests
Tested all client and server launch configs inside and outside of PIE and tested with Launcher.

STRONGLY SUGGESTED: Launch into a deployment in PIE, via receptionist as an external client and via Launcher.

#### Documentation
Documentation linked above, follow up docs for flag setting in progress.

#### Primary reviewers
@m-samiec @girayimprobable

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialGameInstance.h [hunk 1]
```diff
@@ -29,12 +29,14 @@ class SPATIALGDK_API USpatialGameInstance : public UGameInstance
 	// The SpatialWorkerConnection must always be owned by the SpatialGameInstance and so must be created here to prevent TrimMemory from deleting it during Browse.
 	void CreateNewSpatialWorkerConnection();
 
+	// If this flag is set to true standalone clients will not attempt to connect to a deployment automatically if a 'loginToken' exists in arguments.
+	UPROPERTY(Config)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialGameInstance.h]: Is there a reason this is public?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp [hunk 2]
```diff
@@ -165,16 +90,23 @@ void USpatialGameInstance::StartGameInstance()
...
+			if (!LocatorConfig.LoginToken.IsEmpty())
 			{
-				UE_LOG(LogSpatialGameInstance, Fatal, TEXT("Unable to browse to starting map: %s. Application will now exit."), *Error);
-				FPlatformMisc::RequestExit(false);
-			}
+				// Modify the commandline args to have a Host IP to force a NetDriver to be used.
+				const TCHAR* CommandLineArgs = FCommandLine::Get();
 
-			return;
+				FString NewCommandLineArgs = LocatorConfig.LocatorHost + " ";
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialGameInstance.cpp]:
```suggestion
				FString NewCommandLineArgs = LocatorConfig.LocatorHost + TEXT(" ");
```

File: SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/ConnectionConfig.h [hunk 3]
```diff
@@ -61,12 +61,23 @@ struct FConnectionConfig
...
 
-		FParse::Value(CommandLine, TEXT("receptionistHost"), ReceptionistHost);
+		// Parse the commandline for receptionistHost, if it exists then use this as the host IP.
+		if (!FParse::Value(CommandLine, TEXT("receptionistHost"), ReceptionistHost))
+		{
+			// If receptionistHost is not specified then parse for a traditional IP.
+			if (!FParse::Token(CommandLine, ReceptionistHost, 0) || **ReceptionistHost == '-' || **ReceptionistHost == '/')
+			{
+				// If an IP is not specified then use default.
+				UE_LOG(LogTemp, Warning, TEXT("Could not find 'receptionistHost' or IP address in arguments for ReceptionistConfig. Using default: %s"), *SpatialConstants::LOCAL_HOST);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/ConnectionConfig.h]: Please don't use LogTemp. Either create a category, or absorb this log into another suitable one.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/ConnectionConfig.h [hunk 4]
```diff
@@ -61,12 +61,23 @@ struct FConnectionConfig
...
+		: ReceptionistPort(SpatialConstants::DEFAULT_PORT)
 	{
 		const TCHAR* CommandLine = FCommandLine::Get();
 
-		FParse::Value(CommandLine, TEXT("receptionistHost"), ReceptionistHost);
+		// Parse the commandline for receptionistHost, if it exists then use this as the host IP.
+		if (!FParse::Value(CommandLine, TEXT("receptionistHost"), ReceptionistHost))
+		{
+			// If receptionistHost is not specified then parse for a traditional IP.
+			if (!FParse::Token(CommandLine, ReceptionistHost, 0) || **ReceptionistHost == '-' || **ReceptionistHost == '/')
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/ConnectionConfig.h]: Use to array operator to get the first character.

Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/ConnectionConfig.h]: It's called the bracket operator MIKE PLEASE

**Comment by m-samiec** [comment]
Nice clean up.
Could you test connecting in PIE to a external managed server.
And ensure server/client travel still works.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialGameInstance.h (hunk 1 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialGameInstance.h]: Moved to private.

**Pull request merged**